espnet2/cls/espnet_model.py

```diff
@@ -0,0 +1,346 @@
+
+        # for data-parallel
+        speech = speech[:, : speech_lengths.max()]
+
+        if self.frontend is not None:
+            # Frontend
+            #  e.g. STFT and Feature extract
+            #       data_loader may send time-domain signal in this case
+            # speech (Batch, NSamples) -> feats: (Batch, NFrames, Dim)
+            feats, feats_lengths = self.frontend(speech, speech_lengths)
+        else:
+            # No frontend and no feature extract
+            feats, feats_lengths = speech, speech_lengths
+        return feats, feats_lengths
+
+    def setup_metrics_(self):
+        if self.classification_type == "multi-class":
+            return {
+                "acc": EvalFunction.multiclass_accuracy,
+                "macro_precision": partial(
+                    EvalFunction.multiclass_precision,
+                    average="macro",
+                    num_classes=self.vocab_size,
+                ),
+            }
+        elif self.classification_type == "multi-label":
+            return {
+                "acc": partial(EvalFunction.multilabel_accuracy, criteria="hamming"),
+                # acc is usually high if data is imabalanced
+                "mAP": partial(
+                    EvalFunction.multilabel_auprc,
+                    average="macro",
+                    num_labels=self.vocab_size,
+                ),
+            }
+
+
+def label_to_onehot(
+    label: torch.Tensor,
+    label_lengths: torch.Tensor,
+    vocab_size: int,
+    classification_type: str,
+    lsm_weight: float = 0.0,
+) -> torch.Tensor:
+    """Convert label to onehot.
+    Args
+        label: (Batch, Length) pad value should be -1
+        label_lengths: (Batch,) only used in asserts
+        vocab_size: int
+        classification_type: str "multi-class" or "multi-label"
+        lsm_weight: float, label smoothing weight
+    Returns
+        onehot: (Batch, Length, vocab_size)
+    """
+    if classification_type == "multi-class":
+        assert label_lengths.max() == 1, "Only one label per sample"
+        return F.one_hot(label.squeeze(-1), vocab_size).float()
+    elif classification_type == "multi-label":
+        assert (
+            label_lengths.min() == label_lengths.max() or label.min() == -1
+        ), "Pad value should be -1"
+        label = label.masked_fill(label == -1, vocab_size)
+        onehot = F.one_hot(label.view(-1), vocab_size + 1)
+        onehot = onehot[:, :-1]  # Remove dummy column
+        onehot = onehot.view(label.size(0), -1, vocab_size)
+        onehot = onehot.sum(dim=1)
+        onehot = onehot.float()
+        if lsm_weight > 0.0:
+            onehot = onehot * (1 - 2 * lsm_weight) + lsm_weight
+        return onehot
+    else:
+        raise ValueError(
+            "Valid classification types are 'multi-label' and 'multi-class'"
+        )
+
+
+def mixup_augment(speech: torch.Tensor, onehot: torch.Tensor, mixup_prob: float):
```

Comment:
Maybe it is better to move mixup somewhere else in the toolkit. 
Maybe espnet.layers ? 
I know it is most for classification but it may have other uses too (e.g. theoretically one can use it also with HuBERT style pretraining I think).

Reply:
Done, thanks.